providers/implementations/kdfs/hkdf.c.in

```diff
@@ -966,61 +1004,77 @@ static int kdf_tls1_3_derive(void *vctx, unsigned char *key, size_t keylen,
+                          ['KDF_PARAM_KEY',               'key',    'octet_string'],
+                          ['KDF_PARAM_SALT',              'salt',   'octet_string'],
+                          ['KDF_PARAM_FIPS_KEY_CHECK',    'ind_k',  'int'],
+                          ['KDF_PARAM_FIPS_DIGEST_CHECK', 'ind_d',  'int'],
+                          ['KDF_PARAM_PREFIX',            'prefix', 'octet_string'],
+                          ['KDF_PARAM_LABEL',             'label',  'octet_string'],
+                          ['KDF_PARAM_DATA',              'data',   'octet_string'],
+                         )); -}
+
 static int kdf_tls1_3_set_ctx_params(void *vctx, const OSSL_PARAM params[])
 {
-    const OSSL_PARAM *p;
+    struct hkdf_all_set_ctx_params_st p;
     KDF_HKDF *ctx = vctx;
 
-    if (ossl_param_is_empty(params))
-        return 1;
+    if (ctx == NULL || !kdf_tls1_3_set_ctx_params_decoder(params, &p))
+        return 0;
 
-    if (!OSSL_FIPS_IND_SET_CTX_PARAM(ctx, OSSL_FIPS_IND_SETTABLE0, params,
-                                     OSSL_KDF_PARAM_FIPS_DIGEST_CHECK))
+    if (!OSSL_FIPS_IND_SET_CTX_FROM_PARAM(ctx, OSSL_FIPS_IND_SETTABLE0, p.ind_d))
         return 0;
-    if (!OSSL_FIPS_IND_SET_CTX_PARAM(ctx, OSSL_FIPS_IND_SETTABLE1, params,
-                                     OSSL_KDF_PARAM_FIPS_KEY_CHECK))
+    if (!OSSL_FIPS_IND_SET_CTX_FROM_PARAM(ctx, OSSL_FIPS_IND_SETTABLE1, p.ind_k))
         return 0;
 
-    if (!hkdf_common_set_ctx_params(ctx, params))
+    if (!hkdf_common_set_ctx_params(ctx, &p))
         return 0;
 
     if (ctx->mode == EVP_KDF_HKDF_MODE_EXTRACT_AND_EXPAND) {
         ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_MODE);
         return 0;
     }
 
-    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_PREFIX)) != NULL) {
+    if (p.prefix != NULL) {
         OPENSSL_free(ctx->prefix);
         ctx->prefix = NULL;
-        if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->prefix, 0,
+        if (!OSSL_PARAM_get_octet_string(p.prefix, (void **)&ctx->prefix, 0,
                                          &ctx->prefix_len))
             return 0;
     }
 
-    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_LABEL)) != NULL) {
+    if (p.label != NULL) {
         OPENSSL_free(ctx->label);
         ctx->label = NULL;
-        if (!OSSL_PARAM_get_octet_string(p, (void **)&ctx->label, 0,
+        if (!OSSL_PARAM_get_octet_string(p.label, (void **)&ctx->label, 0,
                                          &ctx->label_len))
             return 0;
     }
 
-    OPENSSL_clear_free(ctx->data, ctx->data_len);
-    ctx->data = NULL;
-    if ((p = OSSL_PARAM_locate_const(params, OSSL_KDF_PARAM_DATA)) != NULL
-            && !OSSL_PARAM_get_octet_string(p, (void **)&ctx->data, 0,
-                                            &ctx->data_len))
-        return 0;
+    if (p.data != NULL) {
+        OPENSSL_clear_free(ctx->data, ctx->data_len);
```

Comment:
Hmm this was always clearing data before which seems wrong. Unless it was deliberate..

Reply:
It's cleared in the free too.  I don't enough about the usage of this parameter to know if it needs to be cleared or not.  Suggest that would be a different PR if we want to change it.